woocommerce/woocommerce-ios
# [IAP] Waiting state for purchase and upgrade

Closes: #9878 


## Description


We're adding the ability to purchase the Woo Express Essential plan through In-App Purchase subscriptions.

This PR adds the waiting state we show after the user has authorised the IAP, while we wait for:

1. the payment to be processed, 
2. a success response returned by Apple, 
3. the transaction to be sent to the MobilePay API and verified
4. the store upgrade to complete.

We're reusing the `IndefiniteCircularProgressViewStyle` on the progress view, which has minor edits to make it configurable for our purchases. This is currently only used in the `Preparing reader` step of In Person Payments - it's worth checking that looks correct still (I've checked though!)

### Current plan details

This PR hides the details of the current plan as per the designs, i.e. after the IAP drawer is dismissed and we move on to the next screen. We're not yet in a modal view, so there's still some tasks to make everything match fully, but we're getting closer here.

## Testing instructions


Configure your test device with a sandbox IAP user, in Settings > App Store

1. Open the app and select a Woo Express store in its free trial
2. Tap `Upgrade now` on the banner on the My Store screen
3. Tap `Purchase Debug Essential Monthly`
4. Use your sandbox Apple ID password to complete the IAP
5. Observe that a moment after the IAP drawer is dismissed, the new waiting view is shown, as per designs

N.B. The alert that pops up in the middle is only shown in a Sandbox environment, and we don't have control over it. The `Completed` screen which shows at the end is... not completed. 😉

### Other testing
A note on how we do this... there’s no particularly good signal from Apple for when the In App Purchase drawer closes – the best we have is a `didBecomeActiveNotification` with no distinguishing features in its user info.

We use a closely scoped observation of that notification, and advances the view to `.waiting` when it’s received, as long as the view is still in `.purchasing(_)`, i.e. we’ve not had a cancellation or error or other result from the purchase since we started.

We aggressively stop this observation, because other more concrete results are better indicators, and we don’t need to speculatively advance to the waiting state if we get those. Unfortunately (but understandably, given the lifecycle), the notification usually arrives before the other results, so we delay our handling of the notification by half a second to let everything else catch up.

None of this is ideal, but it works, and it’s what we have.

#### Testing the notification usage
One good example test is to start an IAP, then background the app and foreground it again before the IAP drawer is shown.  You'll see that this notification is received, even though the IAP drawer is then shown on top. Dismissing or completing the IAP will not then trigger this notification again... and therefore won't reach the waiting state.

This is deemed to be uncommon in production; it's much easier in the (slow) Sandbox environment to background the app before the drawer is shown. The underlying code _should_ recover correctly and handle the purchase, even if they don't actually see all the screens we want them to.


## Screenshots


https://github.com/woocommerce/woocommerce-ios/assets/2472348/b81b5411-a5dd-4bd6-b507-1df06954b0c3



---
